dipy/reconst/mcsd.py

```diff
@@ -0,0 +1,347 @@
+        """
+
+        sh_order = response.sh_order
+        super(MultiShellDeconvModel, self).__init__(gtab)
+        B, m, n = multi_tissue_basis(gtab, sh_order, iso)
+
+        delta_f = delta_functions[delta_form]
+        delta = delta_f(response.iso, response.m, response.n, 0., 0.)
+        self.delta = delta
+        multiplier_matrix = _inflate_response(response, gtab, n, delta)
+
+        r, theta, phi = geo.cart2sphere(*reg_sphere.vertices.T)
+        odf_reg, _, _ = shm.real_sym_sh_basis(sh_order, theta, phi)
+        reg = np.zeros([i + iso for i in odf_reg.shape])
+        reg[:iso, :iso] = np.eye(iso)
+        reg[iso:, iso:] = odf_reg
+
+        X = B * multiplier_matrix
+
+        self.fitter = QpFitter(X, reg)
+        self.sh_order = sh_order
+        self._X = X
+        self.sphere = reg_sphere
+        self.response = response
+        self.B_dwi = B
+        self.m = m
+        self.n = n
+
+    def predict(self, params, gtab=None, S0=None):
+        """Compute a signal prediction given spherical harmonic coefficients
+        for the provided GradientTable class instance.
+
+        Parameters
+        ----------
+        params : ndarray
+            The spherical harmonic representation of the FOD from which to make
+            the signal prediction.
+        gtab : GradientTable
+            The gradients for which the signal will be predicted. Use the
+            model's gradient table by default.
+        S0 : ndarray or float
+            The non diffusion-weighted signal value.
+            Default : None
+        """
+        if gtab is None:
+            X = self._X
+        else:
+            iso = self.response.iso
+            B, m, n = multi_tissue_basis(gtab, self.sh_order, iso)
+            multiplier_matrix = _inflate_response(self.response, gtab, n,
+                                                  self.delta)
+            X = B * multiplier_matrix
+        return np.dot(params, X.T)
+
+    @multi_voxel_fit
+    def fit(self, data):
+        coeff = self.fitter(data)
+        return MSDeconvFit(self, coeff, None)
+
+
+class MSDeconvFit(shm.SphHarmFit):
+
+    def __init__(self, model, coeff, mask):
+        self._shm_coef = coeff
+        self.mask = mask
+        self.model = model
+
+    @property
+    def shm_coeff(self):
+        return self._shm_coef[..., self.model.response.iso:]
+
+    @property
+    def volume_fractions(self):
+        tissue_classes = self.model.response.iso + 1
+        return self._shm_coef[..., :tissue_classes] / SH_CONST
+
+
+def _rank(A, tol=1e-8):
+    s = la.svd(A, False, False)
+    threshold = (s[0] * tol)
+    rnk = (s > threshold).sum()
+    return rnk
+
+
+def quadprog(P, Q, G, H):
+    r"""
+    Helper funstion to set up the Quadratic Program (QP) solver in CVXPY.
+    A QP problem has the following form:
+    minimize      1/2 x' P x + Q' x
+    subject to    G x <= H
+
+    Here the QP solver is based on CVXPY and uses OSQP by default.
```

Comment:
Both here and above, the solver is not a default that can be changed. This should be something like "Here the QP solver is based on CVXPY and uses OSQP." without that last phrase.